x/dymns/types/sell_order.go

```diff
@@ -0,0 +1,263 @@
+	if m.HasSetSellPrice() {
+		sellPrice = *m.SellPrice
+	} else {
+		sellPrice = sdk.NewCoin(m.MinPrice.Denom, sdk.ZeroInt())
+	}
+
+	var attrHighestBidder, attrHighestBidPrice sdk.Attribute
+	if m.HighestBid != nil {
+		attrHighestBidder = sdk.NewAttribute(AttributeKeySoHighestBidder, m.HighestBid.Bidder)
+		attrHighestBidPrice = sdk.NewAttribute(AttributeKeySoHighestBidPrice, m.HighestBid.Price.String())
+	} else {
+		attrHighestBidder = sdk.NewAttribute(AttributeKeySoHighestBidder, "")
+		attrHighestBidPrice = sdk.NewAttribute(AttributeKeySoHighestBidPrice, sdk.NewCoin(m.MinPrice.Denom, sdk.ZeroInt()).String())
+	}
+
+	return sdk.NewEvent(
+		EventTypeSellOrder,
+		sdk.NewAttribute(AttributeKeySoAssetId, m.AssetId),
+		sdk.NewAttribute(AttributeKeySoAssetType, m.AssetType.FriendlyString()),
+		sdk.NewAttribute(AttributeKeySoExpiryEpoch, fmt.Sprintf("%d", m.ExpireAt)),
+		sdk.NewAttribute(AttributeKeySoMinPrice, m.MinPrice.String()),
+		sdk.NewAttribute(AttributeKeySoSellPrice, sellPrice.String()),
+		attrHighestBidder,
+		attrHighestBidPrice,
+		sdk.NewAttribute(AttributeKeySoActionName, actionName),
+	)
+}
+
+func (m ActiveSellOrdersExpiration) Validate() error {
+	if len(m.Records) > 0 {
+		uniqueName := make(map[string]bool)
+		// Describe usage of Go Map: only used for validation
+		allNames := make([]string, len(m.Records))
+
+		for i, record := range m.Records {
+			if record.ExpireAt < 1 {
+				return errorsmod.Wrap(gerrc.ErrInvalidArgument, "active SO expiry is empty")
+			}
+
+			if _, duplicated := uniqueName[record.AssetId]; duplicated {
+				return errorsmod.Wrapf(gerrc.ErrInvalidArgument, "active SO is not unique: %s", record.AssetId)
+			}
+
+			uniqueName[record.AssetId] = true
+			allNames[i] = record.AssetId
+		}
+
+		if !sort.StringsAreSorted(allNames) {
+			return errorsmod.Wrap(gerrc.ErrInvalidArgument, "active SO names are not sorted")
+		}
+	}
+
+	return nil
+}
+
+func (m *ActiveSellOrdersExpiration) Sort() {
+	if len(m.Records) < 2 {
+		return
+	}
+
+	sort.Slice(m.Records, func(i, j int) bool {
+		return m.Records[i].AssetId < m.Records[j].AssetId
+	})
+}
+
+func (m *ActiveSellOrdersExpiration) Add(assetId string, expiry int64) {
+	newRecord := ActiveSellOrdersExpirationRecord{AssetId: assetId, ExpireAt: expiry}
+
+	if len(m.Records) < 1 {
+		m.Records = []ActiveSellOrdersExpirationRecord{newRecord}
+		return
+	}
+
+	foundAtIdx := -1
+
+	for i, record := range m.Records {
+		if record.AssetId == assetId {
+			foundAtIdx = i
+			break
+		}
+	}
+
+	if foundAtIdx > -1 {
+		m.Records[foundAtIdx].ExpireAt = expiry
+	} else {
+		m.Records = append(m.Records, newRecord)
+	}
+
+	m.Sort()
+}
+
+func (m *ActiveSellOrdersExpiration) Remove(assetId string) {
+	if len(m.Records) < 1 {
+		return
+	}
+
+	modified := make([]ActiveSellOrdersExpirationRecord, 0, len(m.Records))
+	for _, record := range m.Records {
+		if record.AssetId != assetId {
+			modified = append(modified, record)
+		}
+	}
+	m.Records = modified
+}
```

Comment:
slices.DeleteFunc?
Also there is no reason to return early if len is zero right?

Reply:
optimized, thanks